File: Dynamic Programming/coin change/python/min_coin_change_value.py
# Minimum coins needed for a value
def get_change(m, c):
    n = len(c)
    s = [m]*(m + 1)
    s[0] = 0
    for i in xrange(1, m + 1):
    	for j in xrange(n):
    		if i >= c[j] and s[i - c[j]] + 1 < s[i]:
    			s[i] = s[i - c[j]] + 1
    return s[m]

if __name__ == '__main__':
    m = int(raw_input())
    # c = [1, 3, 4]
    c = map(int, raw_input().split())
    print(get_change(m, c))

File: Dynamic Programming/coin change/cpp/min_coin_change_value.cpp
// Minimum coins needed for a value
#include <bits/stdc++.h>

using namespace std;

int get_change(int m, vector <int> c) {
	vector <int> s(m + 1, m);
	int n = c.size();
	s[0] = 0;
	for (int i = 1; i < m + 1; ++i) {
		for (int j = 0; j < n; ++j) {
			if(i >= c[j] && s[i - c[j]] + 1 < s[i]) {
				s[i] = s[i - c[j]] + 1;
			}
		}
	}
	return s[m];
}

int main() {
	int m, n; cin >> m >> n; // m is the value needed, n is the size of coins array
	vector <int> c(n);
	for (int i = 0; i < n; ++i) {
		cin >> c[i];
	}
	cout << get_change(m, c);
}
